### DISCUSSION TITLE
How can I attach a dependency outside function parameters?

### DISCUSSION THREAD

**Discussion by mths0x5f** [fastapi/fastapi#8274] [Questions] [answered]
**Description**

Is it possible to attach a dependency outside function parameters??

**Additional context**

For example, I want to move this "UserCan" dependency in a decorator somehow, because I would not use the user_id it returns while it must enforce the security policies it implements.

The reason is because linters will warn about unused variables...

```python
@router.get('/{username}', summary='Get user profile details',
            response_model=UserProfileResponse)
async def get_user_profile(username: str,
                           user_id: str = Depends(UserCan('do_ordinary_stuff'))):
```

**Comment by tiangolo** [comment]
I see... 

It is currently not possible, and supporting it in a decorator could lead to some extra complexity that is not really necessary, which linter are you using?

The first thing you can try is just naming the variable `_`. It's common to assign a variable `_` with content that you don't really care. And linters tend to understand that.

So, you could:

```Python
@router.get('/{username}', summary='Get user profile details',
            response_model=UserProfileResponse)
async def get_user_profile(username: str, _ = Depends(UserCan('do_ordinary_stuff'))):
    pass
```

**Comment by mths0x5f** [comment]
@tiangolo it was what I've done the second time I thought about... Thank you for the time took reasoning it

**Comment by dmontagu** [comment]
I ran into a similar issue -- wanting to have security-related dependencies that didn't show up as function arguments. The primary reasons behind this for me were:

1) it seems bad to just let yourself get used to unused argument warnings, and
2) removing what appears to be an unused argument can have *serious* security implications and might seem safe to someone not familiar with FastAPI; I was hoping for a way to make it more obvious what was going on

After some experimentation I came to the following conclusions:
1) I didn't love the solution of using `_` as an argument name since you can't give multiple arguments the same name, but I confirmed that PyCharm doesn't give an unused argument warning for any argument starting with "_". So using `_0`, `_1`, etc. as variable names (or more descriptive alternatives) would accomplish the goal. 
2) If anyone else finds themselves in this position and wants a decorator to add "anonymous" dependencies, the following works for me:
```python
import asyncio
from collections import OrderedDict
from functools import wraps
from inspect import Parameter, signature
from typing import Callable

from fastapi import Depends


def add_dependencies(*dependencies: Depends):
    """
    Adds dependencies to be executed, but not provided as function arguments to the underlying function

    E.g.:
    @add_dependencies(Depends(get_current_active_superuser))
    def update_user(*, user_id: int, user_update: UserUpdate):
        ...
    """

    def decorator(func: Callable):
        sig = signature(func)
        original_parameters = sig.parameters
        new_parameters = OrderedDict(original_parameters)
        for i, dependency in enumerate(dependencies):
            name = _get_anonymous_dependency_name(i, new_parameters)
            new_parameters[name] = Parameter(name, kind=Parameter.KEYWORD_ONLY, default=dependency)

        is_coroutine = asyncio.iscoroutinefunction(func)
        if is_coroutine:
            @wraps(func)
            async def wrapper(*args, **kwargs):
                kwargs = {k: v for k, v in kwargs.items() if k in original_parameters}
                return await func(*args, **kwargs)
        else:
            @wraps(func)
            def wrapper(*args, **kwargs):
                kwargs = {k: v for k, v in kwargs.items() if k in original_parameters}
                return func(*args, **kwargs)

        wrapper.__signature__ = sig.replace(parameters=tuple(new_parameters.values()))
        return wrapper

    return decorator


def _get_anonymous_dependency_name(index: int, current_parameters: OrderedDict):
    name = f"__unnamed_dependency_{index}"
    while name in current_parameters.keys():
        index += 1
        name = f"__unnamed_dependency_{index}"
    return name
```

@tiangolo I'd be happy to submit the above code as a pull request, but would also understand if you'd rather not take on the maintenance burden (in that case I'll just go the underscore-first function argument route).

**Comment by mths0x5f** [comment]
Wow! I am really grateful for your code because I am reluctant of reading FastAPI internals for now

**Comment by mths0x5f** [comment]
I'm reopening the issue for a while as working around with the provided alternative by @dmontagu

**Comment by cjw296** [comment]
Note mainly for myself: @tiangolo tagged this as the issue that will cover "I want all my api endpoints to require authentication by default"

**Comment by tiangolo** [comment, marked as answer]
This is now implemented in version `0.22.0` :tada: 

You can do:

```Python
@app.get("/items/", dependencies=[Depends(verify_token), Depends(verify_key)])
async def read_items():
    return [{"item": "Foo"}, {"item": "Bar"}]
```

The new docs are here: https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-in-path-operation-decorators/

---

And you can also use it in routers, like:

```Python
app.include_router(
    items.router,
    prefix="/items",
    tags=["items"],
    dependencies=[Depends(get_token_header)],
)
```

The relevant docs are here: https://fastapi.tiangolo.com/tutorial/bigger-applications/#include-an-apirouter-with-a-prefix-tags-responses-and-dependencies

(This should solve your use case @cjw296 )

**Comment by mths0x5f** [comment]
Oh, gosh, I love you.

**Comment by tiangolo** [comment]
Hehe :joy: Thanks @mths0x5f ! :smile: :tada: 

I think this is now solved, so I'll close this issue.

But feel free to add more comments, create new issues, etc.

**Comment by tearf001** [comment]
Is there somewhere to get the results of `Depends` registered in APIRouter ?
```py
app.include_router(
    items.router,
    prefix="/items",
    tags=["items"],
    dependencies=[Depends(get_token_header)],
)
```
I am finding ways to get result of  `Depends`  in `router` or `app` scope  other than every specific route.
There a sence, for example, I have a  `Sliding Expiration` route_class to generate a refresh token when the route-class handles every request and found the comming token will soon expires in its sliding-off window 
```py
class SlidingExpirationRoute(APIRoute):
    def __init__(self, *args, **kwargs):
        dependencies = list(kwargs.pop('dependencies', None) or [])
        #... we can also hack some extra depends in ...
        dependencies.insert(0, Depends(get_token_header))
        dependencies.insert(0, Depends(reusable_oauth2))
        # ... more injecting..
        kwargs['dependencies'] = dependencies
        super().__init__(*args, **kwargs)
        
    def get_route_handler(self) -> Callable:
        original_route_handler = super().get_route_handler()

        async def custom_route_handler(request: Request) -> Response:      
            response = await original_route_handler(request)
            # TODO how could I get the results of injected values instead of following hacking
            # hacking
            # these_not_work = await solve_dependencies(request=request, dependant=self.dependant, body=await request.body(),)
            token = await reusable_oauth2(request)
            jwtdata = await get_jwt_data(token)
            exp = datetime.fromtimestamp(jwtdata.exp)
            if exp - datetime.now() < timedelta(seconds=3600):
                newtoken = create_access_token(jwtdata.dict(exclude={'exp'}))
                response.headers['X-refresh-token'] = newtoken
            # end of hacking
            return response
        return custom_route_handler


router = APIRouter(route_class=SlidingExpirationRoute)
@router.get('/path-that-will-refresh-token')
def specific_handler_001():
   pass
```

**Comment by ghandic** [comment]
> I ran into a similar issue -- wanting to have security-related dependencies that didn't show up as function arguments. The primary reasons behind this for me were:
> 
>     1. it seems bad to just let yourself get used to unused argument warnings, and
> 
>     2. removing what appears to be an unused argument can have _serious_ security implications and might seem safe to someone not familiar with FastAPI; I was hoping for a way to make it more obvious what was going on
> 
> 
> After some experimentation I came to the following conclusions:
> 
>     1. I didn't love the solution of using `_` as an argument name since you can't give multiple arguments the same name, but I confirmed that PyCharm doesn't give an unused argument warning for any argument starting with "_". So using `_0`, `_1`, etc. as variable names (or more descriptive alternatives) would accomplish the goal.
> 
>     2. If anyone else finds themselves in this position and wants a decorator to add "anonymous" dependencies, the following works for me:
> 
> 
> ```python
> import asyncio
> from collections import OrderedDict
> from functools import wraps
> from inspect import Parameter, signature
> from typing import Callable
> 
> from fastapi import Depends
> 
> 
> def add_dependencies(*dependencies: Depends):
>     """
>     Adds dependencies to be executed, but not provided as function arguments to the underlying function
> 
>     E.g.:
>     @add_dependencies(Depends(get_current_active_superuser))
>     def update_user(*, user_id: int, user_update: UserUpdate):
>         ...
>     """
> 
>     def decorator(func: Callable):
>         sig = signature(func)
>         original_parameters = sig.parameters
>         new_parameters = OrderedDict(original_parameters)
>         for i, dependency in enumerate(dependencies):
>             name = _get_anonymous_dependency_name(i, new_parameters)
>             new_parameters[name] = Parameter(name, kind=Parameter.KEYWORD_ONLY, default=dependency)
> 
>         is_coroutine = asyncio.iscoroutinefunction(func)
>         if is_coroutine:
>             @wraps(func)
>             async def wrapper(*args, **kwargs):
>                 kwargs = {k: v for k, v in kwargs.items() if k in original_parameters}
>                 return await func(*args, **kwargs)
>         else:
>             @wraps(func)
>             def wrapper(*args, **kwargs):
>                 kwargs = {k: v for k, v in kwargs.items() if k in original_parameters}
>                 return func(*args, **kwargs)
> 
>         wrapper.__signature__ = sig.replace(parameters=tuple(new_parameters.values()))
>         return wrapper
> 
>     return decorator
> 
> 
> def _get_anonymous_dependency_name(index: int, current_parameters: OrderedDict):
>     name = f"__unnamed_dependency_{index}"
>     while name in current_parameters.keys():
>         index += 1
>         name = f"__unnamed_dependency_{index}"
>     return name
> ```
> 
> @tiangolo I'd be happy to submit the above code as a pull request, but would also understand if you'd rather not take on the maintenance burden (in that case I'll just go the underscore-first function argument route).

@dmontagu this is great! Took some of the logic for something else I'm building, VERY helpful :) Thanks!
